### ISSUE TITLE
Transmutation data lost on LAN SMP (and auxilary suggestions)

### ISSUE THREAD

**Issue by Funwayguy** [sinkillerj/ProjectE#780]
As much as I know you're probably going to ignore this issue for the 3rd time, this is a very annoying issue that needs to be fixed. Being one of the pack developers for the "Journey to the Core" mod pack we are getting reports that items stored in the tablet are being lost completely on LAN SMP for those not hosting. Simply telling a user to host a dedicated server is not a sufficient solution to this bug. Not everyone has the resources or luxury of a dedicated server. In the past I've seen this mod have a lot of troubles relating to saving/loading it's data so here's a few suggestion on where to start with fixing/improving it:

- You can use the "PERSISTED_NBT_TAG" section of the players NBT data to store all of yours under a unique NBTTagCompound for your mod. Minecraft will preserve this data for you even through deaths.
- Switch to using IExtendedEnityProperties to manage/pull said data, Forge will even handle most of the save/loading for you. Example setup:
https://github.com/coolAlias/Forge_Tutorials/blob/master/IExtendedEntityPropertiesTutorial.java
- Setup your sync packets to use the read/write methods in the extended properties to save time.
- Switch to UUIDs, you're going to have to do this sooner or later...

The above suggestions should not only fix a lot of the save/load issues the mod tends to experience both on clients and servers but also make things much easier on you in the long run.

**Comment by sinkillerj** [comment]
Things will be improved eventually, however that will be a time consuming process and is currently not the top issue to resolve.

**Comment by williewillus** [comment]
I'll work on this on the side.

**Comment by Kolatra** [comment]
So we should keep this issue open while it is still an issue and it is being worked on.

**Comment by Funwayguy** [comment]
@Kolatra It should be. That's the whole point of an issue 'tracker'

**Comment by Kolatra** [comment]
This issue is still open on #627, any discussion should be over there.

**Comment by williewillus** [comment]
@Funwayguy I had a preliminary version of this implemented today #782, if you want to clone that branch and compile/test it I'd love another tester for it. (note that I'm not done making the method to migrate to the new data format yet)

**Comment by sinkillerj** [comment]
Regardless we should not be storing knowledge in extended entity properties, it is meant for small values such as a single int.

**Comment by williewillus** [comment]
no it isn't....

**Comment by williewillus** [comment]
it stores things in player.dat, where *everything else* (inventory? enderchest?) is stored

**Comment by williewillus** [comment]
http://gyazo.com/76475290fa0fd8a07445a6e86a4fbaa3

**Comment by sinkillerj** [comment]
Theres a difference between storing items, and storing a list of knowledge consisting of potentially almost every item in the game.

**Comment by williewillus** [comment]
??? There is no way that would near that many items. For the tome we still have the full knowledge precached beforehand as a singleton. We aren't changing how much data is stored here, just where it's stored. There is no limit to an NBT Tag Compiund's size. So any notion of "it won't fit" is moot.
This just seems like a solution that might work. These issues have been plaguing us for a long time and no one has taken it upon themselves to find a solution (or they haven't spoken up). Can we test the pull more thoroughly before passing judgment on its strategies?

**Comment by williewillus** [comment]
Also - transmutation knowledge is literally a list of itemstacks. Yes, just like an inventory. There's not a lot of stuff to store.

**Comment by Funwayguy** [comment]
Can I just say that I'm the developer of InfiniteInvo which extends the players inventory to absurdly stupid size and I have successfully been able to store well over 10 million item stacks before Java (not Minecraft mind you) died from lack of memory to 'render' it. The mod hard limits at the integer limit of ~2.14 billion and has nothing to do with the limitation of how much can be stored in the player.dat file so there is really no excuse. The only real limitation is the recursive limit of the NBT format which shouldn't ever be a problem unless you are doing something horribly wrong.

**Comment by williewillus** [comment]
@Funwayguy can you build the latest on my fork williewillus/bag and see if it works in lan? also test it out with old saves and see if they migrate the data. They worked for me but I want some real world tests to back it up :D

**Comment by Funwayguy** [comment]
@williewillus I'm currently working on my own things at the moment but when I have time I'll give it a test. From first glance I can't see anything wrong with it so far. I do have one more suggestions if you're up for it but I'll send them to your fork instead of here.

**Comment by williewillus** [comment]
@Funwayguy the 1st one will probably be a separate refact, wanna work on one thing at a time :)
as for the 2nd, that's probably a good idea, and for the bags too (only sync the one being updated instead of all 15 every time)

**Comment by williewillus** [comment]
Okay so I've implemented fragmented bag data syncing, but I both don't know how to fragment the knowledge syncing in a wise way nor do I think its needed (it was synced per-player in the old system and worked fine).

The thing is, the clientside knowledge list needs to be cleared before reading new packet data or else you get clientside dupe entries in the table(t). However if we split up the data, we need to tell the packet when to clear data and when not to, and if these packets go in the wrong order then it'll go wack.

I also tried to be clever and use a Set<> to store the knowledge, except that that doesn't work with NBTWhitelist clientside, and I don't want to add a bunch of logic and packets just to sync that so something else can sync :p
